### PR TITLE
Added TH96 

### PR DESCRIPTION
**Pull request by nikchi**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -107,3 +107,10 @@ compile_commands.json
 
 # VIA(L) json files that don't belong in QMK repo
 via*.json
+lib/vusb
+lib/printf
+lib/pico-sdk
+lib/lufa
+lib/googletest
+lib/chibios-contrib
+lib/chibios
```
Comment (zvecr) [inline .gitignore:116]:
```suggestion
```

File: keyboards/th96/chconf.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#pragma once
```
Comment (zvecr) [inline keyboards/th96/chconf.h:1]: This should be removed

File: keyboards/th96/config.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+#define ENCODERS_PAD_B { B0 }
+#define ENCODER_RESOLUTION 2
+#endif
+
+#define LED_CAPS_LOCK_PIN C2
+#define LED_NUM_LOCK_PIN C1
+
+#define DEBOUNCE 5
+
+#define STM32_HSECLK 8000000
```
Comment (zvecr) [inline keyboards/th96/config.h:37]:
```suggestion
```

File: keyboards/th96/config.h [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+#define DRIVER_COUNT 2
+#define DRIVER_ADDR_1 0b1010000
+#define DRIVER_ADDR_2 0b1010011
+#define DRIVER_1_LED_TOTAL 59
+#define DRIVER_2_LED_TOTAL 36
+#define RGB_MATRIX_LED_COUNT (DRIVER_1_LED_TOTAL + DRIVER_2_LED_TOTAL)
+
+#ifdef VIA_ENABLE
+#define VIA_QMK_RGBLIGHT_ENABLE
+#endif
```
Comment (zvecr) [inline keyboards/th96/config.h:65]:
```suggestion
```

File: keyboards/th96/config.h [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+// #define ENABLE_RGB_MATRIX_SOLID_REACTIVE_NEXUS
+// #define ENABLE_RGB_MATRIX_SOLID_REACTIVE_MULTINEXUS
+// #define ENABLE_RGB_MATRIX_SPLASH
+// #define ENABLE_RGB_MATRIX_MULTISPLASH
+// #define ENABLE_RGB_MATRIX_SOLID_SPLASH
+// #define ENABLE_RGB_MATRIX_SOLID_MULTISPLASH
+
+
+#define VIAL_KEYBOARD_UID {0x6B, 0x32, 0x00, 0x91, 0x25, 0x8D, 0x36, 0x31}
+
```
Comment (zvecr) [inline keyboards/th96/config.h:120]:
```suggestion
```

File: keyboards/th96/keymap.c [hunk 6]
```diff
@@ -0,0 +1,74 @@
+#include QMK_KEYBOARD_H
```
Comment (zvecr) [inline keyboards/th96/keymap.c:1]: File needs to be removed

File: keyboards/th96/keymaps/vial/1121.vil [hunk 7]
```diff
@@ -0,0 +1 @@
...NO", "KC_NO", "KC_NO"]], [["KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO"], ["RESET", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", -1, "KC_NO"], ["KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO"], ["KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", -1, "KC_NO", "KC_NO", "KC_NO", "KC_NO"], ["KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", -1, "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", -1], ["KC_NO", "KC_NO", "KC_NO", -1, -1, "KC_NO", -1, -1, -1, "KC_NO", "KC_NO", -1, "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO", "KC_NO"]]], "encoder_layout": [[], [], [], []], "layout_options": -1, "macro": [[], [], [], [], [], [], [], [], [], [], [], [], [], [], [], []], "vial_protocol": -1, "via_protocol": 9, "tap_dance": [], "combo": [], "key_override": [], "settings": {}}
```
Comment (zvecr) [inline keyboards/th96/keymaps/vial/1121.vil:1]: Entire vial keymap has to be removed

File: keyboards/th96/mcuconf.h [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+#undef STM32_PLLQ_VALUE
+#undef STM32_PPRE1
+#undef STM32_PPRE2
+
+#define STM32_PLLM_VALUE                    4
+#define STM32_PLLN_VALUE                    168
+#define STM32_PLLP_VALUE                    4
+#define STM32_PLLQ_VALUE                    7
+#define STM32_PPRE1                         STM32_PPRE1_DIV2
+#define STM32_PPRE2                         STM32_PPRE2_DIV1
```
Comment (zvecr) [inline keyboards/th96/mcuconf.h:16]:
```suggestion
```

File: keyboards/th96/rules.mk [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# MCU name
+MCU = STM32F401
+
+# Bootloader selection
+BOOTLOADER = stm32-dfu
+
+STM32_BOOTLOADER_ADDRESS = 0x1FFF0000
+
```
Comment (zvecr) [inline keyboards/th96/rules.mk:8]:
```suggestion

```

File: keyboards/th96/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  if(clockwise){
+    uint16_t keycode = dynamic_keymap_get_keycode(biton32(layer_state),0,17);
+    tap_code(keycode);
+  }
+  else{
+    uint16_t keycode = dynamic_keymap_get_keycode(biton32(layer_state),1,17);
+    tap_code(keycode);
+  }
+  return true;
+}
```
Comment (zvecr) [inline keyboards/th96/keymaps/default/keymap.c:74]: This hack isnt accepted any more, <https://docs.qmk.fm/#/feature_encoders?id=encoder-map>

**Review by zvecr — CHANGES_REQUESTED** [review]
* missing licence headers
* `keyboards/th96/keymaps/default/keymap.c:Zone.Identifier` needs to be removed

File: keyboards/th96/keymaps/default/keymap.c (hunk 10 above)
Comment (nikchi) [reply keyboards/th96/keymaps/default/keymap.c]: removed default keymap

File: keyboards/th96/keymaps/vial/1121.vil (hunk 7 above)
Comment (nikchi) [reply keyboards/th96/keymaps/vial/1121.vil:1]: removed vial keymap

File: keyboards/th96/keymap.c (hunk 6 above)
Comment (nikchi) [reply keyboards/th96/keymap.c]: removed file

File: keyboards/th96/chconf.h (hunk 2 above)
Comment (nikchi) [reply keyboards/th96/chconf.h]: removed

File: keyboards/th96/chconf.h [hunk 11]
```diff
@@ -0,0 +1,17 @@
+#define CH_CFG_ST_FREQUENCY 10000
```
Comment (zvecr) [inline keyboards/th96/chconf.h:1]: This file still needs to be removed, not just the first line removed

File: .gitignore [hunk 12]
```diff
@@ -107,3 +107,4 @@ compile_commands.json
 
 # VIA(L) json files that don't belong in QMK repo
 via*.json
+
```
Comment (zvecr) [inline .gitignore:110]: Shouldnt be any changes to this file
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]
* files are still missing licence headers
  * also missing name/year
* You still need to have a default keymap
* Renaming the vial keymap still means it contains vial content, and needs to be removed....

File: keyboards/th96/config.h [hunk 13]
```diff
@@ -0,0 +1,113 @@
+#pragma once
```
Comment (drashna) [inline keyboards/th96/config.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/th96/config.h [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+/* USB Device descriptor parameter
+#define VENDOR_ID       0x3151
+#define PRODUCT_ID      0x4010
+#define DEVICE_VER      0x0001
+#define PRODUCT         TH96
+#define MANUFACTURER    EPOMAKER
+*/
```
Comment (drashna) [inline keyboards/th96/config.h:9]:
```suggestion
```

File: keyboards/th96/halconf.h [hunk 15]
```diff
@@ -0,0 +1,5 @@
+#pragma once
```
Comment (drashna) [inline keyboards/th96/halconf.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/th96/keymaps/default/config.h [hunk 16]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (drashna) [inline keyboards/th96/keymaps/default/config.h:1]: remove?

File: keyboards/th96/keymaps/default/keymap.c [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+		QK_BOOT,KC_NO,  KC_NO,  KC_NO,    KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,    KC_NO,             KC_NO,    KC_NO,
+		KC_NO,  KC_NO,  KC_NO,  KC_NO,    KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,    KC_NO,             KC_NO,    KC_NO,    KC_NO,    KC_NO,
+		KC_NO,  KC_NO,  KC_NO,  KC_NO,    KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,                       KC_NO,    KC_NO,    KC_NO,    KC_NO,
+		KC_NO,  KC_NO,  KC_NO,  KC_NO,    KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  RGB_VAI,                     KC_NO,    KC_NO,    KC_NO,
+		KC_NO,  KC_NO,  KC_NO,                            KC_NO,                  KC_TRNS,KC_TRNS,RGB_SPI,RGB_VAD,  RGB_SPD,                     KC_NO,    KC_NO,    KC_NO
+    )
+
+};
+
+layer_state_t layer_state_set_user(layer_state_t state) {
```
Comment (drashna) [inline keyboards/th96/keymaps/default/keymap.c:68]: Please convert to GPIO commands: 
https://docs.qmk.fm/#/gpio_control?id=functions

File: keyboards/th96/keymaps/nikchi/config.h [hunk 18]
```diff
@@ -0,0 +1,4 @@
+#define TAPPING_TERM 200
```
Comment (drashna) [inline keyboards/th96/keymaps/nikchi/config.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Also, needs a `#pragma once`

File: keyboards/th96/keymaps/nikchi/keymap.c [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+		KC_NO,   KC_NO,  KC_NO,  KC_NO,   KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,   KC_NO,  KC_NO,              KC_NO,    KC_NO,    KC_NO,    KC_NO,
+		KC_NO,   KC_NO,  KC_NO,  KC_NO,   KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,           KC_NO,              KC_NO,    KC_NO,    KC_NO,    KC_NO,
+		KC_NO,   KC_NO,  KC_NO,  KC_NO,   KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,  KC_NO,   RGB_VAI,                    KC_NO,    KC_NO,    KC_NO,
+		KC_NO,   KC_NO,  KC_NO,                   KC_NO,                          KC_TRNS,KC_TRNS,RGB_SPI, RGB_VAD,RGB_SPD,                      KC_NO,    KC_NO,    KC_NO
+    )
+
+};
+
+layer_state_t layer_state_set_user(layer_state_t state) {
+	palSetPadMode(GPIOC,  0, PAL_MODE_OUTPUT_PUSHPULL);
```
Comment (drashna) [inline keyboards/th96/keymaps/nikchi/keymap.c:113]: same

**Review by drashna — CHANGES_REQUESTED** [review]

**Pull request closed without merging**